### PR TITLE
Bug 1451103 - add method for partition strategy

### PR DESCRIPTION
**Pull request by erindcole**
Wondering if I could get some feedback on this method. It implements a priority queue and groups the data with a list of lists.

### CODE REVIEW

File: moztelemetry/dataset.py [hunk 1]
```diff
@@ -40,6 +41,30 @@ def _group_by_size_greedy(obj_list, tot_groups):
...
     return groups
 
+def _group_by_equal_size(obj_list, tot_groups, threshold=pow(2, 32)):
+    sub_list = []
+    groups = [[]]
+    pq = []
+    tmp_object = none
+
+    if tot_groups <= 1:
+        group_by_size_greedy(obj_list, tot_groups)
```
Comment (acmiyaguchi) [inline moztelemetry/dataset.py]: typo: There should be an underscore prefixing this function

File: moztelemetry/dataset.py [hunk 2]
```diff
@@ -40,6 +41,30 @@ def _group_by_size_greedy(obj_list, tot_groups):
...
+
+    pq = heapq.heapify(obj_list)
```
Comment (acmiyaguchi) [inline moztelemetry/dataset.py]: Items in the priority queue should be tuples of `(priority, object)` as per the [second example](https://docs.python.org/2/library/heapq.html#basic-examples) in the heapq docs.

File: moztelemetry/dataset.py [hunk 3]
```diff
@@ -72,7 +97,7 @@ class Dataset:
         schema = ['submissionDate', 'docType', 'platform']
 
         records = Dataset(bucket, schema) \\
-            .select(
+                .select(
```
Comment (acmiyaguchi) [inline moztelemetry/dataset.py]: I think these are accidental whitespacing changes from a formatter. I would use `git add -p` to manually inspect the changes that you want to make.

File: moztelemetry/dataset.py [hunk 4]
```diff
@@ -40,6 +41,30 @@ def _group_by_size_greedy(obj_list, tot_groups):
...
+    if tot_groups <= 1:
+        group_by_size_greedy(obj_list, tot_groups)
+
+    pq = heapq.heapify(obj_list)
+
+    while pq: 
+        if pq[0].size >= threshold:
+            tmp_object = heapq.heappop(pq)
+            groups.append(tmp_object)
+    else:
```
Comment (acmiyaguchi) [inline moztelemetry/dataset.py]: Typo: proper indentation

File: moztelemetry/dataset.py [hunk 5]
```diff
@@ -40,6 +41,30 @@ def _group_by_size_greedy(obj_list, tot_groups):
...
+        if sub_list.size >= threshold:
```
Comment (acmiyaguchi) [inline moztelemetry/dataset.py]: The `sub_list` doesn't have a `.size` method.

File: moztelemetry/dataset.py (hunk 2 above)
Comment (erindcole) [reply moztelemetry/dataset.py]: I use the heapify() method to make the heap, found [here](https://docs.python.org/2/library/heapq.html#heapq.heapify). It takes a list and sorts it in place. I believe the default behavior is to prioritize by smallest.

File: moztelemetry/dataset.py (hunk 3 above)
Comment (erindcole) [reply moztelemetry/dataset.py]: Okay I will, thanks.

File: moztelemetry/dataset.py (hunk 2 above)
Comment (acmiyaguchi) [reply moztelemetry/dataset.py]: I think heapify needs a comparison operator like `.__cmp__` or `.__lt__` to be defined. Dictionaries are like hash-tables, and don't have a natural order like integers or strings do.

File: moztelemetry/dataset.py (hunk 5 above)
Comment (erindcole) [reply moztelemetry/dataset.py]: Thanks for catching this. I meant to look that up for proper syntax.

**Comment by erindcole** [comment]
Thanks for the comments and feedback. I will fix the smaller errors and will test the method better. How do you suggest I run the unit tests? It seems I almost need to have this method included in the repo first in order to fully test its performance against my Benchmark tests.

**Comment by acmiyaguchi** [comment]
You can add your tests around here.

https://github.com/mozilla/python_moztelemetry/blob/4148c5f64d5ee0e669815922579ac3e87f33bc8e/tests/test_dataset.py#L221-L238

The tests for the new method will probably look similar to this.

**Pull request closed without merging**